hooks/tk-multi-publish2/basic/collector.py

```diff
@@ -0,0 +1,291 @@
+            self.logger.debug("Work file template defined for Mari collection.")
+
+        return channel_item, layer_item
+
+    def _collect_files(self, settings, channel_item, layer_item):
+        """
+        Collect texture files to be published
+        :param dict settings: Configured settings for this collector
+        :param channel_item:   Parent Item instance for channels
+        :param layer_item:   Parent item instance for layers
+        """
+
+        # check that we are currently inside a project:
+        if not mari.projects.current():
+            raise TankError("You must be in an open Mari project to be able to publish!")
+
+        publisher = self.parent
+
+        icon_path = os.path.join(
+            self.disk_location,
+            os.pardir,
+            "icons",
+            "texture.png"
+        )
+
+        # Look for all layers for all channels on all geometry.  Create items for both
+        # the flattened channel as well as the individual layers
+        for geo in mari.geo.list():
+            geo_name = geo.name()
+            
+            for channel in geo.channelList():
+                channel_name = channel.name()
+
+                # find all publishable layers:
+                publishable_layers = self._find_publishable_layers_r(channel.layerList())
+                if not publishable_layers:
+                    # no layers to publish!
+                    continue
+
+                # add item for whole flattened channel:
+                item_name = "%s, %s" % (geo.name(), channel.name())
+                item = channel_item.create_item(
+                    "mari.texture",
+                    "Channel",
+                    item_name
+                )
+                item.set_icon_from_path(icon_path)
+                item.thumbnail_enabled = True
+                
+                item.properties["geo_publish_name"] = geo_name
+                item.properties["channel_publish_name"] = channel_name
+                item.set_thumbnail_from_path(self._extract_mari_thumbnail())
+
+                # add item for each publishable layer:
+                found_layer_names = set()
+                for layer in publishable_layers:
+                    
+                    # for now, duplicate layer names aren't allowed!
+                    layer_name = layer.name()
+                    if layer_name in found_layer_names:
+                        # we might want to handle this one day...
+                        pass
+                    found_layer_names.add(layer_name)
+
+                    item_name = "%s, %s (%s)" % (geo.name(), channel.name(), layer_name)
+                    item = layer_item.create_item(
+                        "mari.texture",
+                        "Layer",
+                        item_name
+                    )
+                    item.set_icon_from_path(icon_path)
+                    item.set_thumbnail_from_path(self._extract_mari_thumbnail())

```

Comment:
Looks like the same thumbnail is generated over and over for each channel/layer. Perhaps this should happen once? Is there a way to set the state of Mari to get a thumbnail with only the channel/layer displayed?